packages/framework/aqueduct/src/data-object-factories/converterDataObjectFactory.ts

```diff
@@ -0,0 +1,193 @@
+	 *     return root.get<IFluidHandle<SharedMap>>("mapKey") !== undefined;
+	 * }
+	 * ```
+	 */
+	isConversionNeeded: (root: ISharedDirectory) => Promise<boolean>;
+
+	/**
+	 * Data required for running conversion. This is necessary because the conversion must happen synchronously.
+	 *
+	 * An example of what to asynchronously retrieve could be getting the "old" DDS that you want to convert the data of:
+	 * ```
+	 * async (root) => {
+	 *     root.get<IFluidHandle<SharedMap>>("mapKey").get();
+	 * }
+	 * ```
+	 */
+	asyncGetDataForConversion: (root: ISharedDirectory) => Promise<TConversionData>;
+
+	/**
+	 * Convert the DataObject upon resolve (i.e. on retrieval of the DataStore).
+	 *
+	 * An example implementation could be changing which underlying DDS is used to represent the DataObject's data:
+	 * ```
+	 * (runtime, root, data) => {
+	 *     // ! These are not all real APIs and are simply used to convey the purpose of this method
+	 *     const mapContent = data.getContent();
+	 *     const newDirectory = SharedDirectory.create(runtime);
+	 *     newDirectory.populateContent(mapContent);
+	 *     root.set("directoryKey", newDirectory.handle);
+	 *     root.delete("mapKey");
+	 * }
+	 * ```
+	 * @param data - Provided by the "asyncGetDataForConversion" function
+	 */
+	convertDataObject: (
+		runtime: FluidDataStoreRuntime,
+		root: ISharedDirectory,
+		data: TConversionData,
+	) => void;
+
+	/**
+	 * If not provided, the Container will be closed after conversion due to underlying changes affecting the data model.
+	 */
+	refreshDataObject?: () => Promise<void>;
+}
+
+/**
+ * ConverterDataObjectFactory is the IFluidDataStoreFactory for converting DataObjects.
+ * See ConverterDataObjectFactoryProps for more information on how to utilize this factory.
+ *
+ * @internal
+ */
+export class ConverterDataObjectFactory<
+	TObj extends DataObject<I>,
+	TConversionData,
+	I extends DataObjectTypes = DataObjectTypes,
+> extends DataObjectFactory<TObj, I> {
+	private convertLock = false;
+
+	// ! TODO: add new DataStoreMessageType.Conversion
+	private static readonly conversionContent = "conversion";
+
+	public constructor(props: ConverterDataObjectFactoryProps<TObj, TConversionData, I>) {
+		const submitConversionOp = (runtime: FluidDataStoreRuntime): void => {
+			runtime.submitMessage(
+				DataStoreMessageType.ChannelOp,
+				ConverterDataObjectFactory.conversionContent,
+				undefined,
+			);
+		};
+
+		const fullConvertDataStore = async (runtime: IFluidDataStoreChannel): Promise<void> => {
+			const realRuntime = runtime as FluidDataStoreRuntime;
+			const root = (await realRuntime.getChannel(
+				dataObjectRootDirectoryId,
+			)) as ISharedDirectory;
+			if (!this.convertLock && (await props.isConversionNeeded(root))) {
+				this.convertLock = true;
+				const data = await props.asyncGetDataForConversion(root);
```

Comment:
Do we have a guarantee that the `data` we're getting is up-to-date? Is this all happening after we've issues a separate barrier op, and the barrier op guarantees that the data here is exactly the state of the DataStore of when the barrier op is in the op-stream and won't be edited after any more?